# Machine Learning With Spark ML
In this lab assignment, you will complete a project by going through the following steps:
1. Get the data.
2. Discover the data to gain insights.
3. Prepare the data for Machine Learning algorithms.
4. Select a model and train it.
5. Fine-tune your model.
6. Present your solution.

As a dataset, we use the California Housing Prices dataset from the StatLib repository. This dataset was based on data from the 1990 California census. The dataset has the following columns
1. `longitude`: a measure of how far west a house is (a higher value is farther west)
2. `latitude`: a measure of how far north a house is (a higher value is farther north)
3. `housing_,median_age`: median age of a house within a block (a lower number is a newer building)
4. `total_rooms`: total number of rooms within a block
5. `total_bedrooms`: total number of bedrooms within a block
6. `population`: total number of people residing within a block
7. `households`: total number of households, a group of people residing within a home unit, for a block
8. `median_income`: median income for households within a block of houses
9. `median_house_value`: median house value for households within a block
10. `ocean_proximity`: location of the house w.r.t ocean/sea

---
# 1. Get the data
Let's start the lab by loading the dataset. The can find the dataset at `data/housing.csv`. To infer column types automatically, when you are reading the file, you need to set `inferSchema` to true. Moreover enable the `header` option to read the columns' name from the file.

In [1]:
// TODO: Replace <FILL IN> with appropriate code

val housing = spark.read.format("csv")
    .option("sep", ",") //comma separator (also is the default)
    .option("inferSchema", "true") //autodetect format
    .option("header", "true") //CSV has headers
    .load("data/housing.csv") //file

housing = [longitude: double, latitude: double ... 8 more fields]


[longitude: double, latitude: double ... 8 more fields]

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

---
# 2. Discover the data to gain insights
Now it is time to take a look at the data. In this step we are going to take a look at the data a few different ways:
* See the schema and dimension of the dataset
* Look at the data itself
* Statistical summary of the attributes
* Breakdown of the data by the categorical attribute variable
* Find the correlation among different attributes
* Make new attributes by combining existing attributes

## 2.1. Schema and dimension
Print the schema of the dataset

In [2]:
// TODO: Replace <FILL IN> with appropriate code

housing.printSchema() //dataset schema

root
 |-- longitude: double (nullable = true)
 |-- latitude: double (nullable = true)
 |-- housing_median_age: double (nullable = true)
 |-- total_rooms: double (nullable = true)
 |-- total_bedrooms: double (nullable = true)
 |-- population: double (nullable = true)
 |-- households: double (nullable = true)
 |-- median_income: double (nullable = true)
 |-- median_house_value: double (nullable = true)
 |-- ocean_proximity: string (nullable = true)



Print the number of records in the dataset.

In [3]:
// TODO: Replace <FILL IN> with appropriate code

housing.count() //dataset size without headers

20640

## 2.2. Look at the data
Print the first five records of the dataset.

In [4]:
// TODO: Replace <FILL IN> with appropriate code

housing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|median_house_value|ocean_proximity|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|          452600.0|       NEAR BAY|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|          358500.0|       NEAR BAY|
|  -122.24|   37.85|              52.0|     1467.0|         190.0|     496.0|     177.0|       7.2574|          352100.0|       NEAR BAY|
|  -122.25|   37.85|              52.0|     1274.0|         235.0|     558.0|     219.0|       5.6431|          341300.0|       NEAR BAY|
|  -122.25|   37.85|              

Print the number of records with population more than 10000.

In [5]:
// TODO: Replace <FILL IN> with appropriate code

housing.where("population > 10000").count()

23

## 2.3. Statistical summary
Print a summary of the table statistics for the attributes `housing_median_age`, `total_rooms`, `median_house_value`, and `population`. You can use the `describe` command.

In [6]:
// TODO: Replace <FILL IN> with appropriate code

housing.select("housing_median_age").describe().show()
housing.select("total_rooms").describe().show()
housing.select("median_house_value").describe().show()
housing.select("population").describe().show()

+-------+------------------+
|summary|housing_median_age|
+-------+------------------+
|  count|             20640|
|   mean|28.639486434108527|
| stddev| 12.58555761211163|
|    min|               1.0|
|    max|              52.0|
+-------+------------------+

+-------+------------------+
|summary|       total_rooms|
+-------+------------------+
|  count|             20640|
|   mean|2635.7630813953488|
| stddev|2181.6152515827944|
|    min|               2.0|
|    max|           39320.0|
+-------+------------------+

+-------+------------------+
|summary|median_house_value|
+-------+------------------+
|  count|             20640|
|   mean|206855.81690891474|
| stddev|115395.61587441359|
|    min|           14999.0|
|    max|          500001.0|
+-------+------------------+

+-------+------------------+
|summary|        population|
+-------+------------------+
|  count|             20640|
|   mean|1425.4767441860465|
| stddev|  1132.46212176534|
|    min|               3.0|
|    max|  

Print the maximum age (`housing_median_age`), the minimum number of rooms (`total_rooms`), and the average of house values (`median_house_value`).

In [7]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.sql.functions._

//housing.<FILL IN>
housing.select("housing_median_age").describe().filter("summary = 'max'").show()
housing.select("total_rooms").describe().filter("summary = 'min'").show()
housing.select("median_house_value").describe().filter("summary = 'mean'").show()

+-------+------------------+
|summary|housing_median_age|
+-------+------------------+
|    max|              52.0|
+-------+------------------+

+-------+-----------+
|summary|total_rooms|
+-------+-----------+
|    min|        2.0|
+-------+-----------+

+-------+------------------+
|summary|median_house_value|
+-------+------------------+
|   mean|206855.81690891474|
+-------+------------------+



## 2.4. Breakdown the data by categorical data
Print the number of houses in different areas (`ocean_proximity`), and sort them in descending order.

In [8]:
// TODO: Replace <FILL IN> with appropriate code

housing.groupBy("ocean_proximity").agg(count("ocean_proximity")).sort(desc("count(ocean_proximity)")).show()

+---------------+----------------------+
|ocean_proximity|count(ocean_proximity)|
+---------------+----------------------+
|      <1H OCEAN|                  9136|
|         INLAND|                  6551|
|     NEAR OCEAN|                  2658|
|       NEAR BAY|                  2290|
|         ISLAND|                     5|
+---------------+----------------------+



Print the average value of the houses (`median_house_value`) in different areas (`ocean_proximity`), and call the new column `avg_value` when print it.

In [9]:
// TODO: Replace <FILL IN> with appropriate code

housing.groupBy("ocean_proximity").agg(avg("median_house_value").as("avg_value")).show()

+---------------+------------------+
|ocean_proximity|         avg_value|
+---------------+------------------+
|         ISLAND|          380440.0|
|     NEAR OCEAN|249433.97742663656|
|       NEAR BAY|259212.31179039303|
|      <1H OCEAN|240084.28546409807|
|         INLAND|124805.39200122119|
+---------------+------------------+



Rewrite the above question in SQL.

In [10]:
// TODO: Replace <FILL IN> with appropriate code

housing.createOrReplaceTempView("df")

spark.sql("""
          SELECT ocean_proximity, AVG(median_house_value)
          FROM df
          GROUP BY ocean_proximity
          """).show()

+---------------+-----------------------+
|ocean_proximity|avg(median_house_value)|
+---------------+-----------------------+
|         ISLAND|               380440.0|
|     NEAR OCEAN|     249433.97742663656|
|       NEAR BAY|     259212.31179039303|
|      <1H OCEAN|     240084.28546409807|
|         INLAND|     124805.39200122119|
+---------------+-----------------------+



## 2.5. Correlation among attributes
Print the correlation among the attributes `housing_median_age`, `total_rooms`, `median_house_value`, and `population`. To do so, first you need to put these attributes into one vector. Then, compute the standard correlation coefficient (Pearson) between every pair of attributes in this new vector. To make a vector of these attributes, you can use the `VectorAssembler` Transformer.

In [11]:
import org.apache.spark.ml.linalg.{Matrix, Vectors}
import org.apache.spark.ml.stat.Correlation
import org.apache.spark.sql.Row

val data = Seq(
  Vectors.sparse(4, Seq((0, 1.0), (3, -2.0))),
  Vectors.dense(4.0, 5.0, 0.0, 3.0),
  Vectors.dense(6.0, 7.0, 0.0, 8.0),
  Vectors.sparse(4, Seq((0, 9.0), (3, 1.0)))
)

val df = data.map(Tuple1.apply).toDF("features")
val Row(coeff1: Matrix) = Correlation.corr(df, "features").head
println("Pearson correlation matrix:\n" + coeff1.toString)



Pearson correlation matrix:
1.0                   0.055641488407465814  NaN  0.4004714203168137  
0.055641488407465814  1.0                   NaN  0.9135958615342522  
NaN                   NaN                   1.0  NaN                 
0.4004714203168137    0.9135958615342522    NaN  1.0                 


data = List((4,[0,3],[1.0,-2.0]), [4.0,5.0,0.0,3.0], [6.0,7.0,0.0,8.0], (4,[0,3],[9.0,1.0]))
df = [features: vector]
coeff1 = 


1.0                   0.055641488407465814  NaN  0.4004714203168137
0.055641488407465814  1.0                   NaN  0.9135958615342522
NaN                   NaN                   1.0  NaN
0.4004714203168137    0.9135958615342522    NaN  1.0


In [12]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.VectorAssembler

val va = new VectorAssembler()
            .setInputCols(Array("housing_median_age","total_rooms","median_house_value","population"))
            .setOutputCol("corr_feature")

val housingAttrs = va.transform(housing).select("corr_feature")

housingAttrs.show(5)

+--------------------+
|        corr_feature|
+--------------------+
|[41.0,880.0,45260...|
|[21.0,7099.0,3585...|
|[52.0,1467.0,3521...|
|[52.0,1274.0,3413...|
|[52.0,1627.0,3422...|
+--------------------+
only showing top 5 rows



va = vecAssembler_dca904b033f4
housingAttrs = [corr_feature: vector]


[corr_feature: vector]

In [13]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.linalg.Matrix
import org.apache.spark.ml.stat.Correlation
import org.apache.spark.sql.Row

val Row(coeff: Matrix) = Correlation.corr(housingAttrs, "corr_feature", "pearson").head

println("Correlation matrix:")

println(s"The standard correlation coefficient:\n ${coeff}")

Correlation matrix:
The standard correlation coefficient:
 1.0                   -0.36126220122231784  0.10562341249318154   -0.2962442397735293   
-0.36126220122231784  1.0                   0.13415311380654338   0.8571259728659772    
0.10562341249318154   0.13415311380654338   1.0                   -0.02464967888891235  
-0.2962442397735293   0.8571259728659772    -0.02464967888891235  1.0                   


coeff = 


1.0                   -0.36126220122231784  0.10562341249318154   -0.2962442397735293
-0.36126220122231784  1.0                   0.13415311380654338   0.8571259728659772
0.10562341249318154   0.13415311380654338   1.0                   -0.02464967888891235
-0.2962442397735293   0.8571259728659772    -0.02464967888891235  1.0


## 2.6. Combine and make new attributes
Now, let's try out various attribute combinations. In the given dataset, the total number of rooms in a block is not very useful, if we don't know how many households there are. What we really want is the number of rooms per household. Similarly, the total number of bedrooms by itself is not very useful, and we want to compare it to the number of rooms. And the population per household seems like also an interesting attribute combination to look at. To do so, add the three new columns to the dataset as below. We will call the new dataset the `housingExtra`.
```
rooms_per_household = total_rooms / households
bedrooms_per_room = total_bedrooms / total_rooms
population_per_household = population / households
```

In [14]:
// TODO: Replace <FILL IN> with appropriate code

val housingCol1 = housing.withColumn("rooms_per_household",($"total_rooms")/($"households"))
val housingCol2 = housingCol1.withColumn("bedrooms_per_room",($"total_bedrooms")/($"total_rooms"))
val housingExtra = housingCol2.withColumn("population_per_household",($"population")/($"households"))

housingExtra.select("rooms_per_household", "bedrooms_per_room", "population_per_household").show(5)

+-------------------+-------------------+------------------------+
|rooms_per_household|  bedrooms_per_room|population_per_household|
+-------------------+-------------------+------------------------+
|  6.984126984126984|0.14659090909090908|      2.5555555555555554|
|  6.238137082601054|0.15579659106916466|       2.109841827768014|
|  8.288135593220339|0.12951601908657123|      2.8022598870056497|
| 5.8173515981735155|0.18445839874411302|       2.547945205479452|
|  6.281853281853282| 0.1720958819913952|      2.1814671814671813|
+-------------------+-------------------+------------------------+
only showing top 5 rows



housingCol1 = [longitude: double, latitude: double ... 9 more fields]
housingCol2 = [longitude: double, latitude: double ... 10 more fields]
housingExtra = [longitude: double, latitude: double ... 11 more fields]


[longitude: double, latitude: double ... 11 more fields]

---
## 3. Prepare the data for Machine Learning algorithms
Before going through the Machine Learning steps, let's first rename the label column from `median_house_value` to `label`.

In [15]:
// TODO: Replace <FILL IN> with appropriate code

val renamedHousing = housingExtra.withColumnRenamed("median_house_value","label")
renamedHousing.show(1)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+
only showing top 1 row



renamedHousing = [longitude: double, latitude: double ... 11 more fields]


[longitude: double, latitude: double ... 11 more fields]

Now, we want to separate the numerical attributes from the categorical attribute (`ocean_proximity`) and keep their column names in two different lists. Moreover, sice we don't want to apply the same transformations to the predictors (features) and the label, we should remove the label attribute from the list of predictors. 

In [16]:
// label columns
val colLabel = "label"

// categorical columns
val colCat = "ocean_proximity"

// numerical columns
val colNum = renamedHousing.columns.filter(_ != colLabel).filter(_ != colCat)

colLabel = label
colCat = ocean_proximity
colNum = Array(longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, rooms_per_household, bedrooms_per_room, population_per_household)


Array(longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, rooms_per_household, bedrooms_per_room, population_per_household)

## 3.1. Prepare continuse attributes
### Data cleaning
Most Machine Learning algorithms cannot work with missing features, so we should take care of them. As a first step, let's find the columns with missing values in the numerical attributes. To do so, we can print the number of missing values of each continues attributes, listed in `colNum`.

In [17]:
// TODO: Replace <FILL IN> with appropriate code
for (c <- colNum) {
    //println(c)
    //println(c.getClass()) //String
    //println(colNum.getClass()) //String
    //println(renamedHousing.select(c).getClass()) // DataFrame
    //renamedHousing.select(c).show(1)
    //println(renamedHousing.col(c).contains(null))
    //renamedHousing.filter(col(c).isNull).show(1)
    //renamedHousing.select(c).groupBy(null).count().show(5)
    //contador = sum(renamedHousing.col(c).isNull.cast("int"))
    println("===========new loop===========")
}


/*======================== PENDING TO DO ========================*/

===========new loop===========
===========new loop===========
===========new loop===========
===========new loop===========
===========new loop===========
===========new loop===========
===========new loop===========
===========new loop===========
===========new loop===========
===========new loop===========
===========new loop===========


As we observerd above, the `total_bedrooms` and `bedrooms_per_room` attributes have some missing values. One way to take care of missing values is to use the `Imputer` Transformer, which completes missing values in a dataset, either using the mean or the median of the columns in which the missing values are located. To use it, you need to create an `Imputer` instance, specifying that you want to replace each attribute's missing values with the "median" of that attribute.

In [18]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.Imputer

val imputer = new Imputer().setStrategy("median")
                            .setInputCols(Array("total_bedrooms","bedrooms_per_room"))
                            .setOutputCols(Array("total_bedrooms","bedrooms_per_room"))
val imputedHousing = imputer.fit(renamedHousing).transform(renamedHousing)

imputedHousing.select("total_bedrooms", "bedrooms_per_room").show(5)

+--------------+-------------------+
|total_bedrooms|  bedrooms_per_room|
+--------------+-------------------+
|         129.0|0.14659090909090908|
|        1106.0|0.15579659106916466|
|         190.0|0.12951601908657123|
|         235.0|0.18445839874411302|
|         280.0| 0.1720958819913952|
+--------------+-------------------+
only showing top 5 rows



imputer = imputer_1f2c61a368e1
imputedHousing = [longitude: double, latitude: double ... 11 more fields]


[longitude: double, latitude: double ... 11 more fields]

### Scaling
One of the most important transformations you need to apply to your data is feature scaling. With few exceptions, Machine Learning algorithms don't perform well when the input numerical attributes have very different scales. This is the case for the housing data: the total number of rooms ranges from about 6 to 39,320, while the median incomes only range from 0 to 15. Note that scaling the label attribues is generally not required.

One way to get all attributes to have the same scale is to use standardization. In standardization, for each value, first it subtracts the mean value (so standardized values always have a zero mean), and then it divides by the variance so that the resulting distribution has unit variance. To do this, we can use the `StandardScaler` Estimator. To use `StandardScaler`, again we need to convert all the numerical attributes into a big vectore of features using `VectorAssembler`, and then call `StandardScaler` on that vactor.

In [19]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.{VectorAssembler, StandardScaler}

val va = new VectorAssembler().setInputCols(Array("total_bedrooms","bedrooms_per_room"))
            .setOutputCol("std_features")
val featuredHousing = va.transform(imputedHousing)

val scaler = new StandardScaler().setWithMean(true).setWithStd(true)
                                .setInputCol("std_features")
                                .setOutputCol("scaled")
val scaledHousing = scaler.fit(featuredHousing).transform(featuredHousing)

scaledHousing.select("scaled").show(5)

/*======================== PENDING TO VERIFY ========================*/

+--------------------+
|              scaled|
+--------------------+
|[-0.9724529206674...|
|[1.35711055309738...|
|[-0.8270042288561...|
|[-0.7197060135854...|
|[-0.6124077983148...|
+--------------------+
only showing top 5 rows



va = vecAssembler_3346beb2307e
featuredHousing = [longitude: double, latitude: double ... 12 more fields]
scaler = stdScal_adba7faa8f0f
scaledHousing = [longitude: double, latitude: double ... 13 more fields]


[longitude: double, latitude: double ... 13 more fields]

## 3.2. Prepare categorical attributes
After imputing and scaling the continuse attributes, we should take care of the categorical attributes. Let's first print the number of distict values of the categirical attribute `ocean_proximity`.

In [20]:
// TODO: Replace <FILL IN> with appropriate code

renamedHousing.select(count("ocean_proximity")).show

+----------------------+
|count(ocean_proximity)|
+----------------------+
|                 20640|
+----------------------+



### String indexer
Most Machine Learning algorithms prefer to work with numbers. So let's convert the categorical attribute `ocean_proximity` to numbers. To do so, we can use the `StringIndexer` that encodes a string column of labels to a column of label indices. The indices are in [0, numLabels), ordered by label frequencies, so the most frequent label gets index 0.

In [21]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.StringIndexer

val indexer = new StringIndexer().setInputCol("ocean_proximity").setOutputCol("idx_ocean_proximity")
val idxHousing = indexer.fit(renamedHousing).transform(renamedHousing)

idxHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+-------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|idx_ocean_proximity|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+-------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|                3.0|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|358500.0|       NEAR BAY|  6.238137082601054|0

indexer = strIdx_0f6bc4a0bce4
idxHousing = [longitude: double, latitude: double ... 12 more fields]


[longitude: double, latitude: double ... 12 more fields]

Now we can use this numerical data in any Machine Learning algorithm. You can look at the mapping that this encoder has learned using the `labels` method: "<1H OCEAN" is mapped to 0, "INLAND" is mapped to 1, etc.

In [22]:
indexer.fit(renamedHousing).labels

Array(<1H OCEAN, INLAND, NEAR OCEAN, NEAR BAY, ISLAND)

### One-hot encoding
Now, convert the label indices built in the last step into one-hot vectors. To do this, you can take advantage of the `OneHotEncoderEstimator` Estimator.

In [23]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.OneHotEncoderEstimator

val encoder = new OneHotEncoderEstimator()
                    .setInputCols(Array("idx_ocean_proximity"))
                    .setOutputCols(Array("oh_idx_ocean_proximity"))
val ohHousing = encoder.fit(idxHousing).transform(idxHousing)

ohHousing.select("oh_idx_ocean_proximity").show(5) 
//this shows a dense array (not sparse) so It means that array is of size 5(from 0 to 4) with 3rd position 1.0

+----------------------+
|oh_idx_ocean_proximity|
+----------------------+
|         (4,[3],[1.0])|
|         (4,[3],[1.0])|
|         (4,[3],[1.0])|
|         (4,[3],[1.0])|
|         (4,[3],[1.0])|
+----------------------+
only showing top 5 rows



encoder = oneHotEncoder_a4e506d3674d
ohHousing = [longitude: double, latitude: double ... 13 more fields]


[longitude: double, latitude: double ... 13 more fields]

---
# 4. Pipeline
As you can see, there are many data transformation steps that need to be executed in the right order. For example, you called the `Imputer`, `VectorAssembler`, and `StandardScaler` from left to right. However, we can use the `Pipeline` class to define a sequence of Transformers/Estimators, and run them in order. A `Pipeline` is an `Estimator`, thus, after a Pipeline's `fit()` method runs, it produces a `PipelineModel`, which is a `Transformer`.

Now, let's create a pipeline called `numPipeline` to call the numerical transformers you built above (`imputer`, `va`, and `scaler`) in the right order from left to right, as well as a pipeline called `catPipeline` to call the categorical transformers (`indexer` and `encoder`). Then, put these two pipelines `numPipeline` and `catPipeline` into one pipeline.

In [24]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.{Pipeline, PipelineModel, PipelineStage}

val numPipeline = "imputer, va, scaler"
val catPipeline = "indexer, encoder"
//val pipeline = new Pipeline().setStages(Array(numPipeline, catPipeline)) // as the prof expected
val pipeline = new Pipeline().setStages(Array(imputer, va, scaler, indexer, encoder))
val newHousing = pipeline.fit(renamedHousing).transform(renamedHousing)

newHousing.show(5)

/*======================== PENDING TO IMPROVE ========================*/

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+-------------------+----------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|        std_features|              scaled|idx_ocean_proximity|oh_idx_ocean_proximity|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+-------------------+----------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908

numPipeline = imputer, va, scaler
catPipeline = indexer, encoder
pipeline = pipeline_b955dc1958d3
newHousing = [longitude: double, latitude: double ... 15 more fields]


[longitude: double, latitude: double ... 15 more fields]

Now, use `VectorAssembler` to put all attributes of the final dataset `newHousing` into a big vector, and call the new column `features`.

In [25]:
// TODO: Replace <FILL IN> with appropriate code

val va2 = new VectorAssembler()
            .setInputCols(Array("longitude", "latitude", "housing_median_age", "total_rooms", "total_bedrooms", 
                                "population", "households", "median_income", "rooms_per_household", 
                                "bedrooms_per_room", "population_per_household","std_features",
                                "scaled", "oh_idx_ocean_proximity"))
            .setOutputCol("features")
val dataset = va2.transform(newHousing).select("features", "label")

dataset.show(5)

/*======================== PENDING TO VERIFY ========================*/

+--------------------+--------+
|            features|   label|
+--------------------+--------+
|[-122.23,37.88,41...|452600.0|
|[-122.22,37.86,21...|358500.0|
|[-122.24,37.85,52...|352100.0|
|[-122.25,37.85,52...|341300.0|
|[-122.25,37.85,52...|342200.0|
+--------------------+--------+
only showing top 5 rows



va2 = vecAssembler_abc5a98fddad
dataset = [features: vector, label: double]


[features: vector, label: double]

---
# 5. Make a model
Here we going to make four different regression models:
* Linear regression model
* Decission tree regression
* Random forest regression
* Gradient-booster forest regression

But, before giving the data to train a Machine Learning model, let's first split the data into training dataset (`trainSet`) with 80% of the whole data, and test dataset (`testSet`) with 20% of it.

In [26]:
// TODO: Replace <FILL IN> with appropriate code

val Array(trainSet, testSet) = dataset.randomSplit(Array(0.8,0.2), 42) // Why 42?
/*======================== PENDING TO CLARIFY ========================*/

trainSet = [features: vector, label: double]
testSet = [features: vector, label: double]


[features: vector, label: double]

## 5.1. Linear regression model
Now, train a Linear Regression model using the `LinearRegression` class. Then, print the coefficients and intercept of the model, as well as the summary of the model over the training set by calling the `summary` method.

In [27]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.LinearRegression

// train the model
val lr = new LinearRegression()
val lrModel = lr.fit(trainSet)
val trainingSummary = lrModel.summary

println(s"Coefficients: <FILL IN>, Intercept: <FILL IN>") // missing to complete fill in
println(s"RMSE: <FILL IN>") // missing to complete fill in 
/*======================== PENDING TO DO ========================*/

Coefficients: <FILL IN>, Intercept: <FILL IN>
RMSE: <FILL IN>


lr = linReg_3dd38d533904
lrModel = linReg_3dd38d533904
trainingSummary = org.apache.spark.ml.regression.LinearRegressionTrainingSummary@3ee602ce


org.apache.spark.ml.regression.LinearRegressionTrainingSummary@3ee602ce

Now, use `RegressionEvaluator` to measure the root-mean-square-erroe (RMSE) of the model on the test dataset.

In [28]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.evaluation.RegressionEvaluator

// make predictions on the test data
val predictions = lrModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error.
val evaluator = new RegressionEvaluator().setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
|104027.60604179045| 85800.0|[-124.3,41.8,19.0...|
| 141831.4282501475|103600.0|[-124.3,41.84,17....|
|178373.72140680766| 79000.0|[-124.27,40.69,36...|
| 191437.4513556054| 90100.0|[-124.19,40.73,21...|
| 93896.07868958544| 74600.0|[-124.19,41.78,15...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 69090.19240689334


predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_3195c4c44ded
rmse = 69090.19240689334


69090.19240689334

## 5.2. Decision tree regression
Repeat what you have done on Regression Model to build a Decision Tree model. Use the `DecisionTreeRegressor` to make a model and then measure its RMSE on the test dataset.

In [29]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.DecisionTreeRegressor
import org.apache.spark.ml.evaluation.RegressionEvaluator

val dt = new DecisionTreeRegressor().setMaxDepth(16)

// train the model
val dtModel = dt.fit(trainSet)

// make predictions on the test data
val predictions = dtModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error
val evaluator = new RegressionEvaluator().setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+----------+--------+--------------------+
|prediction|   label|            features|
+----------+--------+--------------------+
|   71100.0| 85800.0|[-124.3,41.8,19.0...|
|  111300.0|103600.0|[-124.3,41.84,17....|
|   84000.0| 79000.0|[-124.27,40.69,36...|
|  130600.0| 90100.0|[-124.19,40.73,21...|
|   59200.0| 74600.0|[-124.19,41.78,15...|
+----------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 70068.20598191491


dt = dtr_886fdd883d78
dtModel = DecisionTreeRegressionModel (uid=dtr_886fdd883d78) of depth 16 with 13849 nodes
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_bbef176f8f70
rmse = 70068.20598191491


70068.20598191491

## 5.3. Random forest regression
Let's try the test error on a Random Forest Model. Youcan use the `RandomForestRegressor` to make a Random Forest model.

In [30]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.RandomForestRegressor
import org.apache.spark.ml.evaluation.RegressionEvaluator

val rf = new RandomForestRegressor().setMaxDepth(3)
                                      .setNumTrees(20)
                                      .setFeatureSubsetStrategy("auto")

// train the model
val rfModel = rf.fit(trainSet)

// make predictions on the test data
val predictions = rfModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error
val evaluator = new RegressionEvaluator().setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
|187836.08515126884| 85800.0|[-124.3,41.8,19.0...|
|  197257.372904225|103600.0|[-124.3,41.84,17....|
| 189521.1143090006| 79000.0|[-124.27,40.69,36...|
|188304.38563661653| 90100.0|[-124.19,40.73,21...|
| 189521.1143090006| 74600.0|[-124.19,41.78,15...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 74559.54483052266


rf = rfr_e41ef35680bd
rfModel = RandomForestRegressionModel (uid=rfr_e41ef35680bd) with 20 trees
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_f41f7e284411
rmse = 74559.54483052266


74559.54483052266

## 5.4. Gradient-boosted tree regression
Fianlly, we want to build a Gradient-boosted Tree Regression model and test the RMSE of the test data. Use the `GBTRegressor` to build the model.

In [31]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.GBTRegressor
import org.apache.spark.ml.evaluation.RegressionEvaluator

val gb = new GBTRegressor().setMaxDepth(10).setMaxIter(10)

// train the model
val gbModel = gb.fit(trainSet)

// make predictions on the test data
val predictions = gbModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error
val evaluator = new RegressionEvaluator().setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
| 70104.93713013394| 85800.0|[-124.3,41.8,19.0...|
|100658.72113570175|103600.0|[-124.3,41.84,17....|
| 82471.25108030274| 79000.0|[-124.27,40.69,36...|
|60761.285941005524| 90100.0|[-124.19,40.73,21...|
|105895.77823144707| 74600.0|[-124.19,41.78,15...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 57822.69890700471


57822.69890700471

gb = gbtr_adce29d876b3
gbModel = GBTRegressionModel (uid=gbtr_adce29d876b3) with 10 trees
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_69d11bbb5ecf
rmse = 57822.69890700471


---
# 6. Hyperparameter tuning
An important task in Machie Learning is model selection, or using data to find the best model or parameters for a given task. This is also called tuning. Tuning may be done for individual Estimators such as LinearRegression, or for entire Pipelines which include multiple algorithms, featurization, and other steps. Users can tune an entire Pipeline at once, rather than tuning each element in the Pipeline separately. MLlib supports model selection tools, such as `CrossValidator`. These tools require the following items:
* Estimator: algorithm or Pipeline to tune (`setEstimator`)
* Set of ParamMaps: parameters to choose from, sometimes called a "parameter grid" to search over (`setEstimatorParamMaps`)
* Evaluator: metric to measure how well a fitted Model does on held-out test data (`setEvaluator`)

`CrossValidator` begins by splitting the dataset into a set of folds, which are used as separate training and test datasets. For example with `k=3` folds, `CrossValidator` will generate 3 (training, test) dataset pairs, each of which uses 2/3 of the data for training and 1/3 for testing. To evaluate a particular `ParamMap`, `CrossValidator` computes the average evaluation metric for the 3 Models produced by fitting the Estimator on the 3 different (training, test) dataset pairs. After identifying the best `ParamMap`, `CrossValidator` finally re-fits the Estimator using the best ParamMap and the entire dataset.

Below, use the `CrossValidator` to select the best Random Forest model. To do so, you need to define a grid of parameters. Let's say we want to do the search among the different number of trees (1, 5, and 10), and different tree depth (5, 10, and 15).

In [32]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.tuning.ParamGridBuilder
import org.apache.spark.ml.evaluation.RegressionEvaluator
import org.apache.spark.ml.tuning.CrossValidator

val paramGrid = new ParamGridBuilder().addGrid(rf.numTrees, Array(1,5,10))
                                        .addGrid(rf.maxDepth, Array(5, 10, 15))
                                        .build()

val evaluator = new RegressionEvaluator().setLabelCol("label").setPredictionCol("prediction").setMetricName("rmse")
val cv = new CrossValidator().setEstimator(lr)
                                .setEstimatorParamMaps(paramGrid)
                                .setEvaluator(evaluator)
                                .setNumFolds(3)
val cvModel = cv.fit(trainSet)

val predictions = cvModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
|104027.60604179045| 85800.0|[-124.3,41.8,19.0...|
| 141831.4282501475|103600.0|[-124.3,41.84,17....|
|178373.72140680766| 79000.0|[-124.27,40.69,36...|
| 191437.4513556054| 90100.0|[-124.19,40.73,21...|
| 93896.07868958544| 74600.0|[-124.19,41.78,15...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 69090.19240689334


paramGrid = 


Array({
	rfr_e41ef35680bd-maxDepth: 5,
	rfr_e41ef35680bd-numTrees: 1
}, {
	rfr_e41ef35680bd-maxDepth: 5,
	rfr_e41ef35680bd-numTrees: 5
}, {
	rfr_e41ef35680bd-maxDepth: 5,
	rfr_e41ef35680bd-numTrees: 10
}, {
	rfr_e41ef35680bd-maxDepth: 10,
	rfr_e41ef35680bd-numTrees: 1
}, {
	rfr_e41ef35680bd-maxDepth: 10,
	rfr_e41ef35680bd-numTrees: 5
}, {
	rfr_e41ef35680bd-maxDepth: 10,
	rfr_e41ef35680bd-numTrees: 10
}, {
	rfr_e41ef35680bd-maxDepth: 15,
	rfr_e41ef35680bd-numTrees: 1
}, {
	rfr_e41ef35680bd-maxDepth: 15,
	rfr_e41ef35680bd-numTrees: 5
}, {
	rfr_e41ef35680bd-maxDepth: 15,
	rfr_e41e...


---
# 7. Custom transformer
At the end of part two, we added extra columns to the `housing` dataset. Here, we are going to implement a Transformer to do the same task. The Transformer should take the name of two input columns `inputCol1` and `inputCol2`, as well as the name of ouput column `outputCol`. It, then, computes `inputCol1` divided by `inputCol2`, and adds its result as a new column to the dataset. The details of the implemeting a custom Tranfomer is explained [here](https://www.oreilly.com/learning/extend-spark-ml-for-your-own-modeltransformer-types). Please read it before before starting to implement it.

First, define the given parameters of the Transformer and implement a method to validate their schemas (`StructType`).

In [33]:
import org.apache.spark.sql.types.{StructField, StructType, DoubleType}
import org.apache.spark.ml.param.{ParamMap, Param, Params}

trait MyParams extends Params {
    final val inputCol1 = new Param[String](this, "inputCol1", "The input column 1")
    final val inputCol2 = new Param[String](this, "inputCol2", "The input column 2")
    final val outputCol = new Param[String](this, "outputCol", "The output column")
    
  protected def validateAndTransformSchema(schema: StructType): StructType = {
      // <FILL IN>
      // Validates and transforms the input schema
      require(isDefined(inputCol1), s"MyParams requires inputCol1 parameter: $inputCol1")
      require(isDefined(inputCol2), s"MyParams requires inputCol2 parameter: $inputCol2")
      require(isDefined(outputCol), s"MyParams requires outputCol parameter: $outputCol")
      
      // Add the return field
      schema.add(StructField($(outputCol), DoubleType, false))
  }
}


defined trait MyParams


Then, extend the class `Transformer`, and implement its setter functions for the input and output columns, and call then `setInputCol1`, `setInputCol2`, and `setOutputCol`. Morever, you need to override the methods `copy`, `transformSchema`, and the `transform`. The details of what you need to cover in these methods is given [here](https://www.oreilly.com/learning/extend-spark-ml-for-your-own-modeltransformer-types).

In [34]:
import org.apache.spark.ml.util.Identifiable
import org.apache.spark.ml.Transformer
import org.apache.spark.ml.param.{ParamMap, Param, Params}
import org.apache.spark.sql.{DataFrame, Dataset}
import org.apache.spark.sql.types.StructType
import org.apache.spark.sql.functions.{col, udf}

class MyTransformer(override val uid: String) extends Transformer with MyParams {
    def this() = this(Identifiable.randomUID("configurablewordcount"))
    
    //def setInputCol1(value: String): <FILL IN>
    def setInputCol1(value: String): this.type = set(inputCol1, value)
    
    //def setInputCol2(value: String): <FILL IN>
    def setInputCol2(value: String): this.type = set(inputCol2, value)
    
    //def setOutputCol(value: String): <FILL IN>
    def setOutputCol(value: String): this.type = set(outputCol, value)

    //override def copy(extra: ParamMap) <FILL IN>
    override def copy(extra: ParamMap) = {
        defaultCopy(extra)
      }

    //override def transformSchema(schema: StructType) <FILL IN>
    override def transformSchema(schema: StructType): StructType = {
        validateAndTransformSchema(schema)
      }
    
    override def transform(dataset: Dataset[_]): DataFrame = {
        dataset.withColumn($(outputCol), col($(inputCol1)) / col($(inputCol2)))

  }
}

defined class MyTransformer


Now, an instance of `MyTransformer`, and set the input columns `total_rooms` and `households`, and the output column `rooms_per_household` and run it over the `housing` dataset.

In [35]:
val myTransformer = new MyTransformer().setInputCol1("total_rooms").setInputCol2("households").setOutputCol("rooms_per_household")

val myDataset = myTransformer.transform(housing).select("rooms_per_household").show(5)

+-------------------+
|rooms_per_household|
+-------------------+
|  6.984126984126984|
|  6.238137082601054|
|  8.288135593220339|
| 5.8173515981735155|
|  6.281853281853282|
+-------------------+
only showing top 5 rows



myTransformer = configurablewordcount_6f77d4074361


myDataset: Unit = ()


configurablewordcount_6f77d4074361

---
# 8. Custom estimator (predictor)
Now, it's time to implement your own linear regression with gradient descent algorithm as a brand new Estimator. The whole code of the Estimator is given to you, and you do not need to implement anything. It is just a sample that shows how to build a custom Estimator.

The gradient descent update for linear regression is:
$$
w_{i+1} = w_{i} - \alpha_{i} \sum\limits_{j=1}^n (w_i^\top x_j - y_j)x_j
$$

where $i$ is the iteration number of the gradient descent algorithm, and $j$ identifies the observation. Here, $w$ represents an array of weights that is the same size as the array of features and provides a weight for each of the features when finally computing the label prediction in the form:

$$
prediction = w^\top \cdot\ x
$$

where $w$ is the final array of weights computed by the gradient descent, $x$ is the array of features of the observation point and $prediction$ is the label we predict should be associated to this observation.

The given `Helper` class implements the helper methods:
* `dot`: implements the dot product of two vectors and the dot product of a vector and a scalar
* `sum`: implements addition of two vectors
* `fill`: creates a vector of predefined size and initialize it with the predefined value

What you need to do is to implement the methods of the Linear Regresstion class `LR`, which are
* `rmsd`: computes the Root Mean Square Error of a given RDD of tuples of (label, prediction) using the formula:
$$
rmse = \sqrt{\frac{\sum\limits_{i=1}^n (label - prediction)^2}{n}}
$$
* `gradientSummand`: computes the following formula:
$$
gs_{ij} = (w_i^\top x_j - y_j)x_j
$$
* `gradient`: computes the following formula:
$$
gradient = \sum\limits_{j=1}^n gs_{ij}
$$

In [36]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._
import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}

case class Instance(label: Double, features: Vector)

object Helper extends Serializable {
  def dot(v1: Vector, v2: Vector): Double = {
    val m = Matrices.dense(1, v1.size, v1.toArray)
    m.multiply(v2).values(0)
  }

  def dot(v: Vector, s: Double): Vector = {
    val baseArray = v.toArray.map(vi => vi * s)
    Vectors.dense(baseArray)
  }

  def sumVectors(v1: Vector, v2: Vector): Vector = {
    val baseArray = ((v1.toArray) zip (v2.toArray)).map { case (val1, val2) => val1 + val2 }
    Vectors.dense(baseArray)
  }

  def fillVector(size: Int, fillVal: Double): Vector = Vectors.dense(Array.fill[Double](size)(fillVal));
}

defined class Instance
defined object Helper


In [44]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.rdd.RDD
import org.apache.spark.mllib.evaluation.RegressionMetrics

class LR() extends Serializable {
  def calcRMSE(labelsAndPreds: RDD[(Double, Double)]): Double = {
    val regressionMetrics = new RegressionMetrics(labelsAndPreds)
    regressionMetrics.rootMeanSquaredError
  }
  
  def gradientSummand(weights: Vector, lp: Instance): Vector = {
    val mult = (Helper.dot(weights, lp.features) - lp.label)
    val seq = (0 to lp.features.size - 1).map(i => lp.features(i) * mult)
    return Vectors.dense(seq.toArray)
  }
  
  def linregGradientDescent(trainData: RDD[Instance], numIters: Int): (Vector, Array[Double]) = {
    val n = trainData.count()
    val d = trainData.take(1)(0).features.size
    var w = Helper.fillVector(d, 0)
    val alpha = 1.0
    val errorTrain = Array.fill[Double](numIters)(0.0)

    for (i <- 0 until numIters) {
      val labelsAndPredsTrain = trainData.map(lp => (lp.label, Helper.dot(w, lp.features)))
      errorTrain(i) = calcRMSE(labelsAndPredsTrain)

      val gradient = trainData.map(lp => gradientSummand(w, lp)).reduce((v1, v2) => Helper.sumVectors(v1, v2))
      val alpha_i = alpha / (n * scala.math.sqrt(i + 1))
      val wAux = Helper.dot(gradient, (-1) * alpha_i)
      w = Helper.sumVectors(w, wAux)
    }
    (w, errorTrain)
  }
}

defined class LR


In [48]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.{PredictionModel, Predictor}

abstract class MyLinearModel[FeaturesType, Model <: MyLinearModel[FeaturesType, Model]]
  extends PredictionModel[FeaturesType, Model] {
}

class MyLinearModelImpl(override val uid: String, val weights: Vector, val trainingError: Array[Double])
    extends MyLinearModel[Vector, MyLinearModelImpl] {

  override def copy(extra: ParamMap): MyLinearModelImpl = defaultCopy(extra)

  def predict(features: Vector): Double = {
    println("Predicting")
    val prediction = Helper.dot(weights, features)
    prediction
  }
}

defined class MyLinearModel
defined class MyLinearModelImpl


In [49]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._

import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}

abstract class MyLinearRegression[
    FeaturesType,
    Learner <: MyLinearRegression[FeaturesType, Learner, Model],
    Model <: MyLinearModel[FeaturesType, Model]]
  extends Predictor[FeaturesType, Learner, Model] {
}

class MyLinearRegressionImpl(override val uid: String)
    extends MyLinearRegression[Vector, MyLinearRegressionImpl, MyLinearModelImpl] {
  def this() = this(Identifiable.randomUID("linReg"))

  override def copy(extra: ParamMap): MyLinearRegressionImpl = defaultCopy(extra)
  
  def train(dataset: Dataset[_]): MyLinearModelImpl = {
    println("Training")

    val numIters = 10

    val instances: RDD[Instance] = dataset.select(
      col($(labelCol)), col($(featuresCol))).rdd.map {
        case Row(label: Double, features: Vector) =>
          Instance(label, features)
      }

    val (weights, trainingError) = new LR().linregGradientDescent(instances, numIters)

    new MyLinearModelImpl(uid, weights, trainingError)
  }
}

defined class MyLinearRegression
defined class MyLinearRegressionImpl


In [50]:
import org.apache.spark.ml.evaluation.RegressionEvaluator

val lr = new MyLinearRegressionImpl().setLabelCol("label").setFeaturesCol("features")
val model = lr.fit(trainSet)
val predictions = model.transform(trainSet)
predictions.select("prediction", "label", "features").show(5)

// Select (prediction, true label) and compute test error.
val evaluator = new RegressionEvaluator().setLabelCol("label").setPredictionCol("prediction").setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

Training
+--------------------+--------+--------------------+
|          prediction|   label|            features|
+--------------------+--------+--------------------+
|-5.56419203823470...| 94600.0|[-124.35,40.54,52...|
|-6.72941460500383...|111400.0|[-124.26,40.58,52...|
|-4.29168222551613...| 76100.0|[-124.25,40.28,32...|
|-8.20572633385203E73|106700.0|[-124.23,40.54,52...|
|-3.49087099857223...| 50800.0|[-124.23,40.81,52...|
+--------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 1.0913839891711686E74


lr = linReg_f3b15a1364ab
model = linReg_f3b15a1364ab
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_0bf7a3b9ef6f
rmse = 1.0913839891711686E74


1.0913839891711686E74

---
# 9. An End-to-End Classification Test
As the last step, you are given a dataset called `data/ccdefault.csv`. The dataset represents default of credit card clients. It has 30,000 cases and 24 different attributes. More details about the dataset is available at `data/ccdefault.txt`. In this task you should make three models, compare their results and conclude the ideal solution. Here are the suggested steps:
1. Load the data.
2. Carry out some exploratory analyses (e.g., how various features and the target variable are distributed).
3. Train a model to predict the target variable (risk of `default`).
  - Employ three different models (logistic regression, decision tree, and random forest).
  - Compare the models' performances (e.g., AUC).
  - Defend your choice of best model (e.g., what are the strength and weaknesses of each of these models?).
4. What more would you do with this data? Anything to help you devise a better solution?